### DISCUSSION TITLE
Porting the encoding capabilities of OkapiBarcode? (Please see this as an idea only)

### DISCUSSION THREAD

**Discussion by ILoveFOSS** [journeyapps/zxing-android-embedded#626] [Ideas]
Hello!
I did research on available barcode decoding and encoding tools and wasn't able to find a library that does both very well. Zxing is a great decoder but lacking encoding capabilities, while most encoders have no decoding features. As Zxing isn't actively developed any more it is difficult to improve its encoding capabilities. However, an advanced Java based barcode encoding library already exists [https://github.com/woo-j/OkapiBarcode](https://github.com/woo-j/OkapiBarcode). So I was hoping someone with the required skills and spare time could port some of its functionality to this project over the coming months/years?

Unfortunately, I am not capable of it and won't be for the foreseeable future (incompatible career choice). I believe the whole FOSS community could benefit a lot by it though. I created this ticket, so someone will see this eventually and like the idea too and maybe start having a try.

Thank you for reading this and have a great day!

**Comment by rkistner** [comment]
The focus of this library is on decoding, and I'd prefer to keep it that way. Since encoding and decoding are completely different processes, there should be no issue in using different libraries for the two. Are there anything specific problems that would be solved for you by having the encoding functionality in the same library?

Maybe it would help to just add some notes / reference to the Okapi library in the readme?

**Reply by ILoveFOSS** [reply]
ZXing already has encoding capabilities, however it can not encode all barcodes it can decode.

The idea came from using Catima ([https://github.com/TheLastProject/Catima/](https://github.com/TheLastProject/Catima/)), a FOSS version of Stocard or Apple Wallet. For this specific app it would be a major improvement if it could encode all barcodes it can decode. Especially as one specific barcode type is heavily used for vouchers but can't be encoded.

I can see many other use cases too. Barcode generation on an android tablet and directly printing from a portable label printer without need for a stationary device. Printing of invoices according to Austrian (and similar) law with android based (mobile) POS systems as each invoice requires to hold a unique code. etc. etc.

The maintainer of Catima already communicated with the OkapiBarcode team ([https://github.com/woo-j/OkapiBarcode/issues/82](https://github.com/woo-j/OkapiBarcode/issues/82)) and seems like it doesn't support Android. So there is no (known) encoder for Android besides zxing-android-embedded.

Therefore, it would require to either make OkapiBarcode support Android or extend the encoding of zxing-android-embedded by the few barcodes it can't encode yet. I expected the second to be of less work overall.

**Reply by ILoveFOSS** [reply]
I also made a test of ZXing's actual de- & encoding capabilities to have a better understanding of the scope of this task.

ZXing can en- & decode the following barcodes
UPC-A, UPC-E, Code 39, Code 93, Code 128, EAN-8, EAN-13, Codabar, ITF, PDF 417, Aztec, QR Code, Data Matrix

ZXing can decode only these barcodes:
RSS-14 (GS1 Databar), RSS-Expanded (GS1 Databar Expanded)

According to ZXing's GitHub page it also supports these barcodes, but I couldn't confirm from tests:
MaxiCode, UPC/EAN Extension 2/5

In order to reach en- & decoding parity only RSS-14 (GS1 Databar) & RSS-Expanded (GS1 Databar Expanded) would actually need to be ported. MaxiCode & UPC/EAN Extension 2/5 would just be the icing on the cake.

This are the implementations of the 4 barcodes of OkapiBarcode if someone reading this is curious:
-) [GS1 Databar](https://github.com/woo-j/OkapiBarcode/blob/master/src/main/java/uk/org/okapibarcode/backend/DataBar14.java)
-) [GS1 Databar Expanded](https://github.com/woo-j/OkapiBarcode/blob/master/src/main/java/uk/org/okapibarcode/backend/DataBarExpanded.java)
-) [UPC/EAN Extension 2/5](https://github.com/woo-j/OkapiBarcode/blob/master/src/main/java/uk/org/okapibarcode/backend/EanUpcAddOn.java)
-) [MaxiCode](https://github.com/woo-j/OkapiBarcode/blob/master/src/main/java/uk/org/okapibarcode/backend/MaxiCode.java)
